executor/merge_join.go

```diff
@@ -215,196 +269,114 @@ func (t *mergeJoinInnerTable) hasNullInJoinKey(row chunk.Row) bool {
-		}
-	}
 
+	innerIter := e.innerTable.groupRowsIter
+	outerIter := e.outerTable.groupRowsIter
 	for !req.IsFull() {
-		hasMore, err := e.joinToChunk(ctx, req)
-		if err != nil || !hasMore {
-			return err
+		if innerIter.Current() == innerIter.End() {
+			if err := e.innerTable.fetchNextInnerGroup(ctx, e); err != nil {
+				return err
+			}
+			innerIter = e.innerTable.groupRowsIter
 		}
-	}
-	return nil
-}
-
-func (e *MergeJoinExec) joinToChunk(ctx context.Context, chk *chunk.Chunk) (hasMore bool, err error) {
-	for {
-		if e.outerTable.row == e.outerTable.iter.End() {
-			err = e.fetchNextOuterRows(ctx, chk.RequiredRows()-chk.NumRows())
-			if err != nil || e.outerTable.chk.NumRows() == 0 {
-				return false, err
+		if outerIter.Current() == outerIter.End() {
+			if err := e.outerTable.fetchNextOuterGroup(ctx, e, req.RequiredRows()-req.NumRows()); err != nil {
+				return err
+			}
+			outerIter = e.outerTable.groupRowsIter
+			// no more data to execute
+			if outerIter.Current() == outerIter.End() {
+				return nil
 			}
 		}
 
 		cmpResult := -1
-		if e.outerTable.selected[e.outerTable.row.Idx()] && e.innerIter4Row.Len() > 0 {
-			cmpResult, err = e.compare(e.outerTable.row, e.innerIter4Row.Current())
+		if innerIter.Current() != innerIter.End() {
+			cmpResult, err = e.compare(outerIter.Current(), innerIter.Current())
 			if err != nil {
-				return false, err
+				return err
 			}
 		}
 
 		if cmpResult > 0 {
-			if err = e.fetchNextInnerRows(); err != nil {
-				return false, err
-			}
+			innerIter.ReachEnd()
 			continue
 		}
 
 		if cmpResult < 0 {
-			e.joiner.onMissMatch(false, e.outerTable.row, chk)
-			if err != nil {
-				return false, err
-			}
-
-			e.outerTable.row = e.outerTable.iter.Next()
-			e.outerTable.hasMatch = false
-			e.outerTable.hasNull = false
-
-			if chk.IsFull() {
-				return true, nil
+			for row := outerIter.Current(); row != outerIter.End() && !req.IsFull(); row = outerIter.Next() {
+				e.joiner.onMissMatch(false, row, req)
 			}
 			continue
 		}
 
-		matched, isNull, err := e.joiner.tryToMatchInners(e.outerTable.row, e.innerIter4Row, chk)
-		if err != nil {
-			return false, err
-		}
-		e.outerTable.hasMatch = e.outerTable.hasMatch || matched
-		e.outerTable.hasNull = e.outerTable.hasNull || isNull
+		for row := outerIter.Current(); row != outerIter.End() && !req.IsFull(); row = outerIter.Next() {
+			if !e.outerTable.filtersSelected[row.Idx()] {
+				e.joiner.onMissMatch(false, row, req)
+				continue
+			}
 
-		if e.innerIter4Row.Current() == e.innerIter4Row.End() {
-			if !e.outerTable.hasMatch {
-				e.joiner.onMissMatch(e.outerTable.hasNull, e.outerTable.row, chk)
+			matched, isNull, err := e.joiner.tryToMatchInners(row, innerIter, req)
+			if err != nil {
+				return err
 			}
-			e.outerTable.row = e.outerTable.iter.Next()
-			e.outerTable.hasMatch = false
-			e.outerTable.hasNull = false
-			e.innerIter4Row.Begin()
-		}
+			e.hasMatch = e.hasMatch || matched
+			e.hasNull = e.hasNull || isNull
```

Comment:
How about replacing these two fields `e.hashMatch` and `e.hashNull` with two local variables?

Reply:
ditto